### ISSUE TITLE
Stutter on texture load

### ISSUE THREAD

**Issue by Allyriadil** [gamax92/wine-overwatch#14]
Whenever the game loads a texture or animation it stutters. However once its loaded it doesnt happen again usually after 10 minutes of stuttering the game runs smoothly because they are cached. 

Specs : Intel 6700k @ 4.2GHZ
GPU : Nvidia GTX 1080
RAM : 16 GB

Wine prefix is running in 64 bit mode with CSMT enabled.

**Comment by zekesonxx** [comment]
I mean, I'm pretty sure @awesie is the only person involved with this repo who actually understands graphics programming.

There's also a lot going on with Wine, D3D11, and Overwatch in terms of performance. Wine D3D11 doesn't really run great anywhere, or with anything. Blizzard's insane optimization of Overwatch is probably the only reason it's playable through Wine without needing a $2k computer.

**Comment by daniel-j** [comment]
I think it has to do with shader compilation.

Try set env `__GL_SHADER_DISK_CACHE` to `1`

The shader cache resides in `~/.nv/`. This might improve the stuttering when you run the game again.

**Comment by setzer22** [comment]
I've been doing some testing and I can confirm setting __GL_SHADER_DISK_CACHE to 1 makes the texture cache persistent. 

My tests consisted in repeatedly playing the same custom game against bots, restarting the game in-between. Performance was almost stable 100% of the time after 4 or 5 games. As suspected, adding a new hero to the game or changing the map creates more stuttering until the new shaders are loaded. Without that flag set, the loading happens everytime I restart the game since the cache is not persistent.

**Comment by zekesonxx** [comment]
Shader generation stuttering is an issue with rewriting any recentish graphical API into a modern one. You see it in Dolphin, RPCS3, and Wine. Sadly, barring some incredibly dramatic new technology, you're not going to stop seeing it. Dolphin is lucky since they can overcome it via [Ubershaders](https://dolphin-emu.org/blog/2017/07/30/ubershaders/), which are only even possible because the GC's GPU just isn't that complicated, and modern GPUs are bloody powerful.

Shader generation/recompilation (and code compilation in general) is pretty much the opposite of an [embarassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel) problem. The shader really does need to be in-place when the game wants it to be, so the only option is to block (and thus cause stutters) while waiting for the CPU to finish generating it. I don't even want to think of what bullshit would happen if you tried to do async shader compilation.

It's possible that converting SM4 (D3D11 shaders) to SPIR-V (instead of to GLSL), then loading them in OpenGL via [GL_ARB_gl_spirv](https://www.khronos.org/news/press/khronos-releases-opengl-4.6-with-spir-v-support) could reduce the stuttering caused on the driver side. However, this would involve completely rewriting the entire Wine shader stack, and would only even work for people with OpenGL 4.6 support, not to mention throwing any possibility we had of macOS support for OW out of the window.

The way to fix shader generation is to *not* generate them during gameplay, instead pre-generating them. However, OW (like basically every other game) only spits the shaders to the graphical APIs right as or right before it needs them. Which means you're reliant on what shaders you've encountered in the past.

The *real* solution to this is not generating shaders at all. However, given Blizzard's history, I don't see an OpenGL or Vulkan graphical backend for Overwatch anytime soon.

TL;DR: it is a thing, it is known, and it'll get better as the Wine team improves shader generation

**Comment by zekesonxx** [comment]
immediately after finishing that, a thought occurs:

If you wanted to be insane about it, you could build a tool that decrypts OW's data files (or maybe just hooks into OverTool/CASCExplorer, idunno), find all the [0x087 DXBC](https://owdev.wiki/DXBC) files, and somehow make Wine generate and render those shaders so they're all in your GPU driver's shader cache.

I have no idea if that would even work, but if it did: that would eliminate basically all of the driver-side shader generation stutter, which (while I don't have any numbers to back this up) I believe is the vast majority of the stutter caused by shader generation.
